### PR TITLE
feat(fill): allow filling based on the label selector

### LINKED ISSUE #3466: [Feature] Fill input field based on label text
When an input field is properly connected to a label it would be nice to be able to fill the input field based on the label text using `await page.fill('text=Name', 'John Doe');`.

It's often possible to fill the input field based on a CSS selector like `id` or `name`, but making it possible to fill by label text would make it more straight-forward to both read and write the test. Especially if you write tests for screens you did not make yourself.

Examples of cases where it should work:
```
<label for="input1">Name</label>
<input id="input1" value="">
```

```
<label>Name <input value=""></label>

```

### PR DESCRIPTION
**Pull request by dgozman**
This enables filling the input based on the connected label:

```html
<label for=target>Name</label><input id=target>
```

```js
await page.fill('text=Name', 'Alice');
```

Fixes #3466.

### CODE REVIEW

File: src/server/injected/injectedScript.ts [hunk 1]
```diff
@@ -438,27 +438,28 @@ export class InjectedScript {
     return 'done';
   }
 
+  findLabelTarget(element: Element): Element | undefined {
+    if (element.nodeName === 'LABEL') {
+      const forId = element.getAttribute('for');
+      if (forId && element.ownerDocument)
+        return element.ownerDocument.querySelector(`[id="${forId}"]`) || undefined;
```
Comment (JoelEinbinder) [inline src/server/injected/injectedScript.ts:445]:
```suggestion
        return element.ownerDocument.querySelector(`[id="${CSS.escape(forId)}"]`) || undefined;
```

File: src/server/injected/injectedScript.ts [hunk 2]
```diff
@@ -438,27 +438,28 @@ export class InjectedScript {
...
+      return element.querySelector('input,select,textarea,meter,progress') || undefined;
+    }
+    return element;
```
Comment (JoelEinbinder) [inline src/server/injected/injectedScript.ts:448]:
```suggestion
    return element.nodeName === 'LABEL' ? (element as HTMLLabelElement).control || undefined : element;
```

File: src/server/injected/injectedScript.ts [hunk 3]
```diff
@@ -261,10 +261,10 @@ export class InjectedScript {
     return this.pollRaf((progress, continuePolling) => {
       if (node.nodeType !== Node.ELEMENT_NODE)
         return 'error:notelement';
-      const element = node as Element;
-      if (!element.isConnected)
+      const element = this.findLabelTarget(node as Element);
+      if (element && !element.isConnected)
         return 'error:notconnected';
-      if (!this.isVisible(element)) {
+      if (!element || !this.isVisible(element)) {
```
Comment (JoelEinbinder) [inline src/server/injected/injectedScript.ts:267]: This means I can't click a label if it is missing a `for`? But in that case shouldn't we just click the label itself.

File: src/server/injected/injectedScript.ts (hunk 2 above)
Comment (dgozman) [reply src/server/injected/injectedScript.ts]: Wonderful! Done.

**Review by JoelEinbinder — APPROVED** [review]

**Comment by mxschmitt** [comment]
Lets maybe add it to the docs if it was not possible before to fill by label text?

**Comment by dgozman** [comment]
> Lets maybe add it to the docs if it was not possible before to fill by label text?

Good idea, done.

**Pull request merged**